File: dotnet/src/Wata.Sdk/Models/Enums/TransactionStatus.cs
namespace Wata.Sdk.Models;

public readonly struct TransactionStatus : IWataOpenEnum, IEquatable<TransactionStatus>
{
    public TransactionStatus(string value) => Value = value ?? throw new ArgumentNullException(nameof(value));

    public static readonly TransactionStatus Created = new("Created");
    public static readonly TransactionStatus Pending = new("Pending");
    public static readonly TransactionStatus Paid = new("Paid");
    public static readonly TransactionStatus Declined = new("Declined");

    public string Value { get; }

    public override string ToString() => Value;
    public bool Equals(TransactionStatus other) => string.Equals(Value, other.Value, StringComparison.Ordinal);
    public override bool Equals(object? obj) => obj is TransactionStatus other && Equals(other);
    public override int GetHashCode() => Value.GetHashCode(StringComparison.Ordinal);
    public static bool operator ==(TransactionStatus left, TransactionStatus right) => left.Equals(right);
    public static bool operator !=(TransactionStatus left, TransactionStatus right) => !left.Equals(right);
    public static implicit operator string(TransactionStatus value) => value.Value;
    public static explicit operator TransactionStatus(string value) => new(value);
}

File: java/src/main/java/pro/wata/sdk/model/TransactionStatus.java
package pro.wata.sdk.model;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

import java.util.Map;
import java.util.Objects;

/** Статус транзакции: {@code Created}, {@code Pending}, {@code Paid}, {@code Declined}. */
public final class TransactionStatus {

    private static final Map<String, TransactionStatus> KNOWN = EnumSupport.registry();

    public static final TransactionStatus CREATED = register("Created");
    public static final TransactionStatus PENDING = register("Pending");
    public static final TransactionStatus PAID = register("Paid");
    public static final TransactionStatus DECLINED = register("Declined");

    private final String value;
    private final boolean known;

    private TransactionStatus(String value, boolean known) {
        this.value = value;
        this.known = known;
    }

    private static TransactionStatus register(String value) {
        TransactionStatus status = new TransactionStatus(value, true);
        KNOWN.put(value, status);
        return status;
    }

    @JsonCreator
    public static TransactionStatus of(String value) {
        return EnumSupport.resolve(value, KNOWN, TransactionStatus::new);
    }

    @JsonValue
    public String value() {
        return value;
    }

    public boolean isKnown() {
        return known;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof TransactionStatus other)) return false;
        return value.equals(other.value);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(value);
    }

    @Override
    public String toString() {
        return value;
    }
}
